File: packages/dart/lib/src/validators/curp_validator.dart
/// CURP (Clave Única de Registro de Población) Validator and Generator
///
/// CURP is Mexico's unique population registry code, issued by RENAPO.
///
/// Structure (18 characters):
/// - 4 letters from name (like RFC)
/// - 6 digits for birth date (YYMMDD)
/// - 1 letter for gender (H=Male, M=Female)
/// - 2 letters for birth state code
/// - 3 consonants from paterno, materno, nombre
/// - 1 alphanumeric differentiator (assigned by RENAPO)
/// - 1 numeric check digit
///
/// Example: OEAF771012HMCRGR09
library;

import 'package:diacritic/diacritic.dart';
import 'package:catalogmx/src/utils/date_utils.dart';
import 'package:catalogmx/src/utils/text_utils.dart';

/// Exception for CURP validation errors
class CURPException implements Exception {
  final String message;
  CURPException(this.message);

  @override
  String toString() => 'CURPException: $message';
}

/// CURP Validator
class CURPValidator {
  final String curp;

  static final RegExp _generalRegex = RegExp(
    r'^[A-Z][AEIOUX][A-Z]{2}[0-9]{2}[0-1][0-9][0-3][0-9][MH][A-Z]{2}[BCDFGHJKLMNPQRSTVWXYZ]{3}[0-9A-Z]{2}$',
  );

  static const int length = 18;

  static const Map<String, String> _stateCodes = {
    'AGUASCALIENTES': 'AS',
    'BAJA CALIFORNIA': 'BC',
    'BAJA CALIFORNIA SUR': 'BS',
    'CAMPECHE': 'CC',
    'COAHUILA': 'CL',
    'COLIMA': 'CM',
    'CHIAPAS': 'CS',
    'CHIHUAHUA': 'CH',
    'CIUDAD DE MEXICO': 'DF',
    'DISTRITO FEDERAL': 'DF',
    'CDMX': 'DF',
    'DURANGO': 'DG',
    'GUANAJUATO': 'GT',
    'GUERRERO': 'GR',
    'HIDALGO': 'HG',
    'JALISCO': 'JC',
    'ESTADO DE MEXICO': 'MC',
    'MEXICO': 'MC',
    'MICHOACAN': 'MN',
    'MORELOS': 'MS',
    'NAYARIT': 'NT',
    'NUEVO LEON': 'NL',
    'OAXACA': 'OC',
    'PUEBLA': 'PL',
    'QUERETARO': 'QT',
    'QUINTANA ROO': 'QR',
    'SAN LUIS POTOSI': 'SP',
    'SINALOA': 'SL',
    'SONORA': 'SR',
    'TABASCO': 'TC',
    'TAMAULIPAS': 'TS',
    'TLAXCALA': 'TL',
    'VERACRUZ': 'VZ',
    'YUCATAN': 'YN',
    'ZACATECAS': 'ZS',
    'NACIDO EN EL EXTRANJERO': 'NE',
    'EXTRANJERO': 'NE',
  };

  static const List<String> _cacophonic = [
    'BACA', 'BAKA', 'BUEI', 'BUEY', 'CACA', 'CACO', 'CAGA', 'CAGO',
    'CAKA', 'KAKO', 'COGE', 'COGI', 'COJA', 'COJE', 'COJI', 'COJO',
    'COLA', 'CULO', 'FALO', 'FETO', 'GETA', 'GUEI', 'GUEY', 'JETA',
    'JOTO', 'KACA', 'KACO', 'KAGA', 'KAGO', 'KOGE', 'KOGI', 'KOJA',
    'KOJE', 'KOJI', 'KOJO', 'KOLA', 'KULO', 'LILO', 'LOCA', 'LOCO',
    'LOKA', 'LOKO', 'MAME', 'MAMO', 'MEAR', 'MEAS', 'MEON', 'MIAR',
    'MION', 'MOCO', 'MOKO', 'MULA', 'MULO', 'NACA', 'NACO', 'PEDA',
    'PEDO', 'PENE', 'PIPI', 'PITO', 'POPO', 'PUTA', 'PUTO', 'QULO',
    'RATA', 'ROBA', 'ROBE', 'ROBO', 'RUIN', 'SENO', 'TETA', 'VACA',
    'VAGA', 'VAGO', 'VAKA', 'VUEI', 'VUEY', 'WUEI', 'WUEY',
  ];

  CURPValidator(String curp) : curp = curp.toUpperCase().trim();

  /// Validates if the CURP is structurally valid
  bool validate() {
    if (curp.length != length) {
      throw CURPException('CURP length must be $length');
    }
    if (!_generalRegex.hasMatch(curp)) {
      throw CURPException('Invalid CURP structure');
    }
    return true;
  }

  /// Alias for validate() that doesn't throw exceptions
  bool isValid() {
    try {
      return validate();
    } catch (e) {
      return false;
    }
  }

  /// Validates the check digit
  bool validateCheckDigit() {
    if (curp.length != 18) return false;

    final curp17 = curp.substring(0, 17);
    final expectedDigit = CURPGenerator.calculateCheckDigit(curp17);
    final actualDigit = curp[17];

    return expectedDigit == actualDigit;
  }

  /// Extracts birth date from CURP
  DateTime? getBirthDate() {
    if (!isValid()) return null;
    return parseDateYYMMDD(curp.substring(4, 10));
  }

  /// Extracts gender from CURP
  String? getGender() {
    if (!isValid()) return null;
    final code = curp[10];
    return code == 'H' ? 'Hombre' : 'Mujer';
  }

  /// Extracts state code from CURP
  String? getStateCode() {
    if (!isValid()) return null;
    return curp.substring(11, 13);
  }
}

/// CURP Generator
class CURPGenerator {
  final String nombre;
  final String apellidoPaterno;
  final String? apellidoMaterno;
  final DateTime fechaNacimiento;
  final String sexo;
  final String? estado;

  static const String _vowels = 'AEIOU';
  static const String _consonants = 'BCDFGHJKLMNPQRSTVWXYZ';

  static const List<String> _excludedWords = [
    'DE', 'LA', 'LAS', 'MC', 'VON', 'DEL', 'LOS', 'Y', 'MAC', 'VAN', 'MI',
    'DA', 'DAS', 'DE', 'DEL', 'DER', 'DI', 'DIE', 'DD', 'EL', 'LA', 'LOS',
    'LAS', 'LE', 'LES', 'MAC', 'MC', 'VAN', 'VON', 'Y',
  ];

  static const List<String> _allowedChars = [
    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
    'N', 'Ñ', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
  ];

  CURPGenerator({
    required this.nombre,
    required this.apellidoPaterno,
    this.apellidoMaterno,
    required this.fechaNacimiento,
    required this.sexo,
    this.estado,
  }) {
    if (apellidoPaterno.trim().isEmpty) {
      throw CURPException('Apellido paterno is required');
    }
    if (nombre.trim().isEmpty) {
      throw CURPException('Nombre is required');
    }
    if (sexo.toUpperCase() != 'H' && sexo.toUpperCase() != 'M') {
      throw CURPException('sexo must be "H" (Hombre) or "M" (Mujer)');
    }
  }

  /// Generates the CURP
  String generate() {
    final letters = _generateLetters();
    final date = _generateDate();
    final gender = sexo.toUpperCase();
    final stateCode = _getStateCode(estado);
    final consonants = _generateConsonants();
    final homoclave = _generateHomoclave();

    return letters + date + gender + stateCode + consonants + homoclave;
  }

  String _cleanName(String name) {
    if (name.isEmpty) return '';
    return cleanName(name, _excludedWords, _allowedChars);
  }

  String get _paternoCalculo => _cleanName(apellidoPaterno);
  String get _maternoCalculo => apellidoMaterno != null ? _cleanName(apellidoMaterno!) : '';
  String get _nombreCalculo => _cleanName(nombre);

  String get _nombreIniciales {
    if (_nombreCalculo.isEmpty) return _nombreCalculo;

    final words = _nombreCalculo.split(' ');
    if (words.length > 1) {
      if (['MARIA', 'JOSE', 'MA', 'MA.', 'J', 'J.'].contains(words[0])) {
        return words.sublist(1).join(' ');
      }
    }
    return _nombreCalculo;
  }

  String _generateLetters() {
    final clave = StringBuffer();
    final paterno = _paternoCalculo;

    if (paterno.isEmpty) {
      throw CURPException('Apellido paterno cannot be empty');
    }

    // First letter of paterno
    clave.write(paterno[0]);

    // First vowel of paterno (after first letter)
    bool vowelFound = false;
    for (var i = 1; i < paterno.length; i++) {
      if (_vowels.contains(paterno[i])) {
        clave.write(paterno[i]);
        vowelFound = true;
        break;
      }
    }
    if (!vowelFound) {
      clave.write('X');
    }

    // First letter of materno (or X if none)
    final materno = _maternoCalculo;
    if (materno.isNotEmpty) {
      clave.write(materno[0]);
    } else {
      clave.write('X');
    }

    // First letter of nombre
    final nombreInit = _nombreIniciales;
    if (nombreInit.isEmpty) {
      throw CURPException('Nombre cannot be empty');
    }
    clave.write(nombreInit[0]);

    String result = clave.toString();

    // Check for cacophonic words - replace second character with 'X'
    if (CURPValidator._cacophonic.contains(result)) {
      result = result[0] + 'X' + result.substring(2);
    }

    return result;
  }

  String _generateDate() {
    return dateToYYMMDD(fechaNacimiento);
  }

  String _getStateCode(String? state) {
    if (state == null || state.isEmpty) return 'NE';

    final stateUpper = state.toUpperCase().trim();

    // Try exact match
    if (CURPValidator._stateCodes.containsKey(stateUpper)) {
      return CURPValidator._stateCodes[stateUpper]!;
    }

    // Try partial match
    for (final entry in CURPValidator._stateCodes.entries) {
      if (entry.key.contains(stateUpper) || stateUpper.contains(entry.key)) {
        return entry.value;
      }
    }

    // If it's already a 2-letter code, return it
    if (stateUpper.length == 2) {
      return stateUpper;
    }

    return 'NE'; // Default to born abroad
  }

  String? _getFirstInternalConsonant(String word) {
    if (word.length <= 1) return 'X';

    for (var i = 1; i < word.length; i++) {
      if (_consonants.contains(word[i])) {
        return word[i];
      }
    }
    return 'X';
  }

  String _generateConsonants() {
    final consonants = StringBuffer();

    // First internal consonant of paterno
    consonants.write(_getFirstInternalConsonant(_paternoCalculo) ?? 'X');

    // First internal consonant of materno
    final materno = _maternoCalculo;
    if (materno.isNotEmpty) {
      consonants.write(_getFirstInternalConsonant(materno) ?? 'X');
    } else {
      consonants.write('X');
    }

    // First internal consonant of nombre
    consonants.write(_getFirstInternalConsonant(_nombreIniciales) ?? 'X');

    return consonants.toString();
  }

  String _generateHomoclave() {
    // Position 17: Differentiator (assigned by RENAPO, we use default)
    final differentiator = fechaNacimiento.year < 2000 ? '0' : 'A';

    // Position 18: Check digit (calculable)
    final tempCurp = _generateLetters() +
        _generateDate() +
        sexo.toUpperCase() +
        _getStateCode(estado) +
        _generateConsonants() +
        differentiator;

    final checkDigit = calculateCheckDigit(tempCurp);

    return differentiator + checkDigit;
  }

  /// Calculates the check digit for a 17-character CURP
  static String calculateCheckDigit(String curp17) {
    if (curp17.length != 17) {
      throw CURPException('CURP must have exactly 17 characters for check digit calculation');
    }

    const dictionary = '0123456789ABCDEFGHIJKLMNÑOPQRSTUVWXYZ';

    int suma = 0;
    for (var i = 0; i < curp17.length; i++) {
      final char = curp17[i];
      int charValue;

      try {
        charValue = dictionary.indexOf(char);
        if (charValue == -1) charValue = 0;
      } catch (e) {
        charValue = 0;
      }

      suma += charValue * (18 - i);
    }

    int digito = 10 - (suma % 10);

    if (digito == 10) {
      digito = 0;
    }

    return digito.toString();
  }
}

// Helper functions

/// Validates a CURP
bool validateCURP(String curp) {
  try {
    final validator = CURPValidator(curp);
    return validator.validate();
  } catch (e) {
    return false;
  }
}

/// Generates a CURP
String generateCURP({
  required String nombre,
  required String apellidoPaterno,
  String? apellidoMaterno,
  required DateTime fechaNacimiento,
  required String sexo,
  String? estado,
}) {
  final generator = CURPGenerator(
    nombre: nombre,
    apellidoPaterno: apellidoPaterno,
    apellidoMaterno: apellidoMaterno,
    fechaNacimiento: fechaNacimiento,
    sexo: sexo,
    estado: estado,
  );
  return generator.generate();
}

File: packages/kotlin/src/main/kotlin/com/openbancor/catalogmx/validators/CurpValidator.kt
package com.openbancor.catalogmx.validators

import com.openbancor.catalogmx.utils.DateUtils
import com.openbancor.catalogmx.utils.TextUtils
import java.time.LocalDate

/**
 * CURP (Clave Única de Registro de Población) Exception
 */
class CurpException(message: String) : Exception(message)

/**
 * CURP (Clave Única de Registro de Población) Validator
 *
 * CURP is Mexico's unique population registry code, issued by RENAPO.
 *
 * Structure (18 characters):
 * - 4 letters from name (like RFC)
 * - 6 digits for birth date (YYMMDD)
 * - 1 letter for gender (H=Male, M=Female)
 * - 2 letters for birth state code
 * - 3 consonants from paterno, materno, nombre
 * - 1 alphanumeric differentiator (assigned by RENAPO)
 * - 1 numeric check digit
 *
 * Example: OEAF771012HMCRGR09
 */
class CurpValidator(curp: String) {
    val curp: String = curp.uppercase().trim()

    companion object {
        private val generalRegex = Regex(
            "^[A-Z][AEIOUX][A-Z]{2}[0-9]{2}[0-1][0-9][0-3][0-9][MH][A-Z]{2}[BCDFGHJKLMNPQRSTVWXYZ]{3}[0-9A-Z]{2}$"
        )

        const val LENGTH = 18

        val stateCodes = mapOf(
            "AGUASCALIENTES" to "AS", "BAJA CALIFORNIA" to "BC",
            "BAJA CALIFORNIA SUR" to "BS", "CAMPECHE" to "CC",
            "COAHUILA" to "CL", "COLIMA" to "CM", "CHIAPAS" to "CS",
            "CHIHUAHUA" to "CH", "CIUDAD DE MEXICO" to "DF",
            "DISTRITO FEDERAL" to "DF", "CDMX" to "DF", "DURANGO" to "DG",
            "GUANAJUATO" to "GT", "GUERRERO" to "GR", "HIDALGO" to "HG",
            "JALISCO" to "JC", "ESTADO DE MEXICO" to "MC", "MEXICO" to "MC",
            "MICHOACAN" to "MN", "MORELOS" to "MS", "NAYARIT" to "NT",
            "NUEVO LEON" to "NL", "OAXACA" to "OC", "PUEBLA" to "PL",
            "QUERETARO" to "QT", "QUINTANA ROO" to "QR",
            "SAN LUIS POTOSI" to "SP", "SINALOA" to "SL", "SONORA" to "SR",
            "TABASCO" to "TC", "TAMAULIPAS" to "TS", "TLAXCALA" to "TL",
            "VERACRUZ" to "VZ", "YUCATAN" to "YN", "ZACATECAS" to "ZS",
            "NACIDO EN EL EXTRANJERO" to "NE", "EXTRANJERO" to "NE"
        )

        val cacophonic = listOf(
            "BACA", "BAKA", "BUEI", "BUEY", "CACA", "CACO", "CAGA", "CAGO",
            "CAKA", "KAKO", "COGE", "COGI", "COJA", "COJE", "COJI", "COJO",
            "COLA", "CULO", "FALO", "FETO", "GETA", "GUEI", "GUEY", "JETA",
            "JOTO", "KACA", "KACO", "KAGA", "KAGO", "KOGE", "KOGI", "KOJA",
            "KOJE", "KOJI", "KOJO", "KOLA", "KULO", "LILO", "LOCA", "LOCO",
            "LOKA", "LOKO", "MAME", "MAMO", "MEAR", "MEAS", "MEON", "MIAR",
            "MION", "MOCO", "MOKO", "MULA", "MULO", "NACA", "NACO", "PEDA",
            "PEDO", "PENE", "PIPI", "PITO", "POPO", "PUTA", "PUTO", "QULO",
            "RATA", "ROBA", "ROBE", "ROBO", "RUIN", "SENO", "TETA", "VACA",
            "VAGA", "VAGO", "VAKA", "VUEI", "VUEY", "WUEI", "WUEY"
        )
    }

    /**
     * Validates if the CURP is structurally valid (throws exception on failure)
     */
    fun validate(): Boolean {
        if (curp.length != LENGTH) {
            throw CurpException("CURP length must be $LENGTH")
        }
        if (!generalRegex.matches(curp)) {
            throw CurpException("Invalid CURP structure")
        }
        return true
    }

    /**
     * Checks if CURP is valid without raising exceptions
     */
    fun isValid(): Boolean {
        return try {
            validate()
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Validates the check digit
     */
    fun validateCheckDigit(): Boolean {
        if (curp.length != 18) return false

        val curp17 = curp.substring(0, 17)
        val expectedDigit = CurpGenerator.calculateCheckDigit(curp17)
        val actualDigit = curp[17]

        return expectedDigit == actualDigit
    }

    /**
     * Extracts birth date from CURP
     */
    fun getBirthDate(): LocalDate? {
        if (!isValid()) return null
        return DateUtils.parseDateYYMMDD(curp.substring(4, 10))
    }

    /**
     * Extracts gender from CURP
     */
    fun getGender(): String? {
        if (!isValid()) return null
        return if (curp[10] == 'H') "Hombre" else "Mujer"
    }

    /**
     * Extracts state code from CURP
     */
    fun getStateCode(): String? {
        if (!isValid()) return null
        return curp.substring(11, 13)
    }
}

/**
 * CURP Generator
 */
class CurpGenerator(
    val nombre: String,
    val apellidoPaterno: String,
    val apellidoMaterno: String? = null,
    val fechaNacimiento: LocalDate,
    val sexo: String,
    val estado: String? = null
) {
    init {
        require(apellidoPaterno.trim().isNotEmpty()) { "Apellido paterno is required" }
        require(nombre.trim().isNotEmpty()) { "Nombre is required" }
        require(sexo.uppercase() == "H" || sexo.uppercase() == "M") { "Sexo must be H or M" }
    }

    companion object {
        private const val VOWELS = "AEIOU"
        private const val CONSONANTS = "BCDFGHJKLMNPQRSTVWXYZ"

        private val excludedWords = listOf(
            "DE", "LA", "LAS", "MC", "VON", "DEL", "LOS", "Y", "MAC", "VAN", "MI",
            "DA", "DAS", "DER", "DI", "DIE", "DD", "EL", "LE", "LES"
        )

        private val allowedChars = listOf(
            'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N',
            'Ñ', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'
        )

        /**
         * Calculates the check digit for a 17-character CURP
         */
        @JvmStatic
        fun calculateCheckDigit(curp17: String): Char {
            if (curp17.length != 17) {
                throw CurpException("CURP must have exactly 17 characters for check digit calculation")
            }

            val dictionary = "0123456789ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"

            var suma = 0
            for (i in curp17.indices) {
                val char = curp17[i]
                val charValue = dictionary.indexOf(char).coerceAtLeast(0)
                suma += charValue * (18 - i)
            }

            var digito = 10 - (suma % 10)
            if (digito == 10) {
                digito = 0
            }

            return digito.toString()[0]
        }
    }

    fun generate(): String {
        val letters = generateLetters()
        val date = generateDate()
        val gender = sexo.uppercase()
        val stateCode = getStateCode(estado)
        val consonants = generateConsonants()
        val homoclave = generateHomoclave()

        return letters + date + gender + stateCode + consonants + homoclave
    }

    private fun cleanName(name: String): String {
        if (name.isEmpty()) return ""
        return TextUtils.cleanName(name, excludedWords, allowedChars)
    }

    private val paternoCalculo: String get() = cleanName(apellidoPaterno)
    private val maternoCalculo: String get() = apellidoMaterno?.let { cleanName(it) } ?: ""
    private val nombreCalculo: String get() = cleanName(nombre)

    private val nombreIniciales: String
        get() {
            if (nombreCalculo.isEmpty()) return nombreCalculo

            val words = nombreCalculo.split(' ')
            if (words.size > 1) {
                if (words[0] in listOf("MARIA", "JOSE", "MA", "MA.", "J", "J.")) {
                    return words.drop(1).joinToString(" ")
                }
            }
            return nombreCalculo
        }

    private fun generateLetters(): String {
        val clave = StringBuilder()
        val paterno = paternoCalculo

        if (paterno.isEmpty()) {
            throw CurpException("Apellido paterno cannot be empty")
        }

        // First letter of paterno
        clave.append(paterno[0])

        // First vowel of paterno (after first letter)
        var vowelFound = false
        for (i in 1 until paterno.length) {
            if (paterno[i] in VOWELS) {
                clave.append(paterno[i])
                vowelFound = true
                break
            }
        }
        if (!vowelFound) {
            clave.append('X')
        }

        // First letter of materno (or X if none)
        val materno = maternoCalculo
        if (materno.isNotEmpty()) {
            clave.append(materno[0])
        } else {
            clave.append('X')
        }

        // First letter of nombre
        val nombreInit = nombreIniciales
        if (nombreInit.isEmpty()) {
            throw CurpException("Nombre cannot be empty")
        }
        clave.append(nombreInit[0])

        var result = clave.toString()

        // Check for cacophonic words - replace second character with 'X'
        if (result in CurpValidator.cacophonic) {
            result = "${result[0]}X${result.substring(2)}"
        }

        return result
    }

    private fun generateDate(): String = DateUtils.dateToYYMMDD(fechaNacimiento)

    private fun getStateCode(state: String?): String {
        if (state.isNullOrEmpty()) return "NE"

        val stateUpper = state.uppercase().trim()

        // Try exact match
        if (stateUpper in CurpValidator.stateCodes) {
            return CurpValidator.stateCodes[stateUpper]!!
        }

        // Try partial match
        for ((key, value) in CurpValidator.stateCodes) {
            if (key.contains(stateUpper) || stateUpper.contains(key)) {
                return value
            }
        }

        // If it's already a 2-letter code, return it
        if (stateUpper.length == 2) {
            return stateUpper
        }

        return "NE" // Default to born abroad
    }

    private fun getFirstInternalConsonant(word: String): Char {
        if (word.length <= 1) return 'X'

        for (i in 1 until word.length) {
            if (word[i] in CONSONANTS) {
                return word[i]
            }
        }
        return 'X'
    }

    private fun generateConsonants(): String {
        val consonants = StringBuilder()

        // First internal consonant of paterno
        consonants.append(getFirstInternalConsonant(paternoCalculo))

        // First internal consonant of materno
        val materno = maternoCalculo
        if (materno.isNotEmpty()) {
            consonants.append(getFirstInternalConsonant(materno))
        } else {
            consonants.append('X')
        }

        // First internal consonant of nombre
        consonants.append(getFirstInternalConsonant(nombreIniciales))

        return consonants.toString()
    }

    private fun generateHomoclave(): String {
        // Position 17: Differentiator (assigned by RENAPO, we use default)
        val differentiator = if (fechaNacimiento.year < 2000) '0' else 'A'

        // Position 18: Check digit (calculable)
        val tempCurp = generateLetters() +
            generateDate() +
            sexo.uppercase() +
            getStateCode(estado) +
            generateConsonants() +
            differentiator

        val checkDigit = calculateCheckDigit(tempCurp)

        return "$differentiator$checkDigit"
    }
}
